flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/org/apache/flink/cdc/connectors/mysql/source/MySqlEventDeserializer.java

```diff
@@ -146,8 +148,11 @@ protected boolean isSchemaChangeRecord(SourceRecord record) {
 
     @Override
     protected TableId getTableId(SourceRecord record) {
-        String[] parts = record.topic().split("\\.");
-        return TableId.tableId(parts[1], parts[2]);
+        Struct value = (Struct) record.value();
+        Struct source = value.getStruct(Envelope.FieldName.SOURCE);
+        String dbName = source.getString(DATABASE_NAME_KEY);
+        String tableName = source.getString(TABLE_NAME_KEY);
+        return TableId.tableId(dbName, tableName);
```

Comment:
Copy from SourceRecordUtils#getTableId.
There may be issues when the table name contains special characters in record.topic() method, for example, `sqlInjection; DROP TABLE important_data;` will be converted to `sqlInjection__DROP_TABLE_important_data_`.